### PR TITLE
Update all benchmarks except thread_pool

### PR DESCRIPTION
**Pull request by Marwes**
Found out that the benchmarks had bitrotted in #1710 so this updates them to use criterion instead of the unstable test crate, letting them run without needing nightly rust.

Also enables benchmark testing on CI.

### CODE REVIEW

File: tokio/benches/tcp.rs [hunk 1]
```diff
@@ -182,76 +196,82 @@ mod transfer {
...
         }
     }
 }
+
+use prelude::*;
+
+fn bench_tcp(c: &mut Criterion) {
+    c.bench_function("connect_churn/one_thread", connect_churn::one_thread);
+    c.bench_function("connect_churn/two_threads", connect_churn::two_threads);
+    c.bench_function("connect_churn/multi_threads", connect_churn::multi_threads);
```
Comment (hawkw) [inline tokio/benches/tcp.rs]: I think this could maybe be converted to a Criterion parameterized benchmark parameterized over the number of threads, but it's fine to do that in a follow-up.

File: tokio/benches/mpsc.rs [hunk 2]
```diff
@@ -268,3 +247,37 @@ fn contended_bounded_tx(b: &mut Bencher) {
...
+    c.bench_function("send_one_message", send_one_message);
+    c.bench_function("send_one_message_large", send_one_message_large);
+    c.bench_function("bounded_rx_not_ready", bounded_rx_not_ready);
+    c.bench_function("bounded_tx_poll_ready", bounded_tx_poll_ready);
+    c.bench_function("bounded_tx_poll_not_ready", bounded_tx_poll_not_ready);
+    c.bench_function("unbounded_rx_not_ready", unbounded_rx_not_ready);
+    c.bench_function("unbounded_rx_not_ready_x5", unbounded_rx_not_ready_x5);
+    c.bench_function("bounded_uncontended_1", bounded_uncontended_1);
+    c.bench_function("bounded_uncontended_1_large", bounded_uncontended_1_large);
+    c.bench_function("bounded_uncontended_2", bounded_uncontended_2);
```
Comment (hawkw) [inline tokio/benches/mpsc.rs]: Some of these benchmarks with different bounds seem like they could maybe be converted into parameterized benchmarks later. But, I think that can be a follow-up.

File: Cargo.toml [hunk 3]
```diff
@@ -11,3 +11,4 @@ members = [
   "examples",
   "tests-integration",
 ]
+
```
Comment (hawkw) [inline Cargo.toml]: unrelated whitespace change?

**Comment by Marwes** [comment]
Loom seems to be timing out, probably unrelated?

**Comment by carllerche** [comment]
hmm, I will have to investigate.

**Comment by carllerche** [comment]
Yes, we need to speed up the loom build...

**Comment by Marwes** [comment]
Rebased with tests passing.

**Comment by carllerche** [comment]
So sorry for the delay, the focus was on getting v0.2 out of the door. Now I can circle back.

Is there a way to avoid having to list all the benches explicitly in `Cargo.toml`? I haven't thought this through, but what if we made a new workspace crate for benches that had a `main` function and used criterion directly?

**Comment by Marwes** [comment]
That would be possible but personally I don't think it is a good idea as it means it won't be possible to compile and run individual benchmark targets with `--bench tcp` etc.

**Comment by carllerche** [comment]
They can still be split up... I guess it is fine to keep them as "benches"...

**Comment by dekellum** [comment]
Any updates on this?  Should this be revived?

**Comment by Marwes** [comment]
Benches were updated and put in their own crate in another PR

**Comment by dekellum** [comment]
What crate? What PR?
